patches/v8/0011-Implement-PromiseContextScope-and-promise_context_ta.patch

```diff
@@ -0,0 +1,323 @@
+Subject: Implement PromiseContextScope and promise_context_tag
+
+Allows optional detection of promises created and waited on within
+different context scopes.
+---
+ include/v8-isolate.h                        | 17 +++++++++++++++++
+ src/api/api.cc                              | 12 ++++++++++++
+ src/builtins/promise-abstract-operations.tq |  4 ++++
+ src/builtins/promise-constructor.tq         |  4 ++++
+ src/builtins/promise-misc.tq                |  9 +++++++++
+ src/common/message-template.h               |  7 ++++++-
+ src/compiler/js-create-lowering.cc          |  4 +++-
+ src/execution/isolate.cc                    |  9 ++++++++-
+ src/execution/isolate.h                     |  5 +++++
+ src/heap/factory.cc                         |  1 +
+ src/objects/js-promise.tq                   |  1 +
+ src/runtime/runtime-promise.cc              | 19 +++++++++++++++++++
+ src/runtime/runtime.h                       |  4 +++-
+ 13 files changed, 92 insertions(+), 4 deletions(-)
+
+diff --git a/include/v8-isolate.h b/include/v8-isolate.h
+index 1d079ac0bb1dd3921d680bcd1958da4bbc114e9e..f4fbe23e4ac6e333719a7a519a46ed184d36318d 100644
+--- a/include/v8-isolate.h
++++ b/include/v8-isolate.h
+@@ -1635,6 +1635,8 @@ class V8_EXPORT Isolate {
+    */
+   void LocaleConfigurationChangeNotification();
+ 
++  class PromiseContextScope;
++
+   Isolate() = delete;
+   ~Isolate() = delete;
+   Isolate(const Isolate&) = delete;
+@@ -1678,6 +1680,21 @@ MaybeLocal<T> Isolate::GetDataFromSnapshotOnce(size_t index) {
+   return Local<T>::FromSlot(slot);
+ }
+ 
++class Isolate::PromiseContextScope {
++public:
++  explicit PromiseContextScope(Isolate* v8_isolate, int32_t tag);
++  PromiseContextScope(const PromiseContextScope&) = delete;
++  PromiseContextScope(PromiseContextScope&&) = delete;
++  PromiseContextScope& operator=(const PromiseContextScope&) = delete;
++  PromiseContextScope& operator=(PromiseContextScope&&) = delete;
++  ~PromiseContextScope();
++
++private:
++  internal::Isolate* i_isolate_;
++  int32_t current_tag_;
++  int32_t previous_tag_ = 0;
++};
++
+ }  // namespace v8
+ 
+ #endif  // INCLUDE_V8_ISOLATE_H_
+diff --git a/src/api/api.cc b/src/api/api.cc
+index bddda91dbb07500eac71274fe859603a416821b8..90d76b2f46398fd501532f5d43598536590b69a3 100644
+--- a/src/api/api.cc
++++ b/src/api/api.cc
+@@ -11610,6 +11610,18 @@ std::string SourceLocation::ToString() const {
+   return std::string(function_) + "@" + file_ + ":" + std::to_string(line_);
+ }
+ 
++Isolate::PromiseContextScope::PromiseContextScope(Isolate* v8_isolate, int32_t tag)
++    : i_isolate_(reinterpret_cast<i::Isolate*>(v8_isolate)),
++      current_tag_(tag),
++      previous_tag_(i_isolate_->promise_context_tag()) {
++  DCHECK_EQ(previous_tag_, 0);
++  i_isolate_->set_promise_context_tag(tag);
++}
++
++Isolate::PromiseContextScope::~PromiseContextScope() {
++  DCHECK_EQ(i_isolate_->promise_context_tag(), (int32_t)current_tag_);  i_isolate_->set_promise_context_tag(previous_tag_);
```

Comment:
Note to self: will need to update this to change the `DCHECK_EQ` to `CHECK_EQ` or remove `current_tag_` to avoid a build error in release mode due to unused `current_tag_`... also need to fix the line wrap